src/kernel/panic.zig

```diff
@@ -1,13 +1,148 @@
+
+    pub fn init(allocator: *std.mem.Allocator) SymbolMap {
+        return SymbolMap{
+            .symbols = ArrayList(MapEntry).init(allocator),
+        };
+    }
+
+    pub fn add(self: *SymbolMap, name: []const u8, addr: u32) !void {
+        try self.symbols.append(MapEntry{ .addr = addr, .func_name = name });
+    }
+
+    pub fn search(self: *const SymbolMap, addr: usize) ?[]const u8 {
+        if (self.symbols.count() == 0)
+            return null;
+        // Find the first element whose address is greater than addr
+        var previous_name: ?[]const u8 = null;
+        for (self.symbols.toSlice()[0..]) |entry| {
+            if (entry.addr > addr)
+                return previous_name;
+            previous_name = entry.func_name;
+        }
+        return previous_name;
+    }
+};
+
+pub fn init(mem_profile: *const mem.MemProfile, allocator: *std.mem.Allocator) !void {
+    log.logInfo("Init panic\n");
+    // Exit if we haven't loaded all debug modules
+    if (mem_profile.boot_modules.len < 1)
+        return;
+    var kmap_start: u32 = 0;
+    var kmap_end: u32 = 0;
+    for (mem_profile.boot_modules) |module| {
+        const mod_start = mem.physToVirt(module.mod_start);
+        const mod_end = mem.physToVirt(module.mod_end);
+        const mod_str_ptr = mem.physToVirt(@intToPtr([*]u8, module.cmdline));
+        const mod_str = mod_str_ptr[0..strlen(mod_str_ptr)];
+        if (std.mem.eql(u8, mod_str, "kernel.map")) {
+            kmap_start = mod_start;
+            kmap_end = mod_end;
+            break;
+        }
+    }
+    std.debug.assert(kmap_start != 0 and kmap_end != 0);
+
+    var syms = SymbolMap.init(allocator);
+    var i: u32 = kmap_start;
+    while (i < kmap_end) {
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        // Parse the address from the first column
+        const addr_start = i;
+        // Each column is separated by one or more whitespace characters, so read until the first one
+        while (!std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        // Convert to an integer
+        const addr_str = @intToPtr([*]u8, addr_start)[0..(i - addr_start)];
+        const addr = try std.fmt.parseInt(u32, addr_str, 16);
+        // Skip to the function name in the second column
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        const name_start = i;
+        while (!std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+        const name = @intToPtr([*]u8, name_start)[0..(i - name_start)];
+        try syms.add(name, addr);
+        // Skip to the next line
+        while (std.fmt.isWhiteSpace(@intToPtr(*u8, i).*)) : (i += 1) {}
+    }
+    symbol_map = syms;
+    log.logInfo("Done\n");
+}
+
+fn strlen(chars: [*]u8) usize {
+    var i: usize = 0;
+    while (chars[i] != 0) : (i += 1) {}
+    return i;
+}
+
+fn logTraceAddress(addr: usize) void {
+    var str: []const u8 = "(no symbols available)"[0..];
```

Comment:
Cheeky one liner ;)
```Zig
const str = if (symbol_map) |syms| syms.search(addr) orelse "?????" else "(no symbols available)";
log.logError("{x}: {}\n", addr, str);
```
This removed the type and `[0..]`

Reply:
Oh nice, does that work? I didn't try that as I thought the compiler would complain that the string literals were of different sizes.